core-api/src/main/java/com/optimizely/ab/bucketing/DecisionService.java

```diff
@@ -198,31 +202,125 @@ public DecisionResponse<FeatureDecision> getVariationForFeature(@Nonnull Feature
                                                                     @Nonnull ProjectConfig projectConfig,
                                                                     @Nonnull List<OptimizelyDecideOption> options) {
-        DecisionReasons reasons = DefaultDecisionReasons.newInstance();
+        return getVariationsForFeatureList(Arrays.asList(featureFlag), user, projectConfig, options).get(0);
+    }
 
-        DecisionResponse<FeatureDecision> decisionVariationResponse = getVariationFromExperiment(projectConfig, featureFlag, user, options);
-        reasons.merge(decisionVariationResponse.getReasons());
+    private UserProfile getUserProfile(String userId, DecisionReasons reasons) {
+        UserProfile userProfile = null;
 
-        FeatureDecision decision = decisionVariationResponse.getResult();
-        if (decision != null) {
-            return new DecisionResponse(decision, reasons);
+        try {
+            Map<String, Object> userProfileMap = userProfileService.lookup(userId);
+            if (userProfileMap == null) {
+                String message = reasons.addInfo("We were unable to get a user profile map from the UserProfileService.");
+                logger.info(message);
+            } else if (UserProfileUtils.isValidUserProfileMap(userProfileMap)) {
+                userProfile = UserProfileUtils.convertMapToUserProfile(userProfileMap);
+            } else {
+                String message = reasons.addInfo("The UserProfileService returned an invalid map.");
+                logger.warn(message);
+            }
+        } catch (Exception exception) {
+            String message = reasons.addInfo(exception.getMessage());
+            logger.error(message);
+            errorHandler.handleError(new OptimizelyRuntimeException(exception));
         }
 
-        DecisionResponse<FeatureDecision> decisionFeatureResponse = getVariationForFeatureInRollout(featureFlag, user, projectConfig);
-        reasons.merge(decisionFeatureResponse.getReasons());
-        decision = decisionFeatureResponse.getResult();
+        if (userProfile == null) {
+            userProfile = new UserProfile(userId, new HashMap<String, Decision>());
+        }
 
-        String message;
-        if (decision.variation == null) {
-            message = reasons.addInfo("The user \"%s\" was not bucketed into a rollout for feature flag \"%s\".",
-                user.getUserId(), featureFlag.getKey());
-        } else {
-            message = reasons.addInfo("The user \"%s\" was bucketed into a rollout for feature flag \"%s\".",
-                user.getUserId(), featureFlag.getKey());
+        return  userProfile;
+    }
+
+    static class UserProfileTracker {
+        public UserProfile userProfile;
+        public boolean profileUpdated;
+
+        UserProfileTracker(UserProfile userProfile, boolean profileUpdated) {
+            this.userProfile = userProfile;
```

Comment:
injected UserProfile can't be null? Can you add @Nonnull notation?

Reply:
refactored UserProfileTracker

Reply:
@raju-opti Can we make UPSTracker as a separate file?

Reply:
Moved to a separate file